### PR TITLE
Add Request Redirects tool that can re-route wp_get_remote calls

### PR DESCRIPTION
**Pull request by moon0326**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds a new tool to the beta tester that can re-route outgoing requests via `wp_get_remote` function.
This tool is designed to test the remote inbox notification on staging.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

### Setup

1. Go to https://staging.woo.com admin (find the username and password from secret store 5926)
9. Go to Ads and publish a new inbox notification.
10. Confirm the added notification from https://staging.woocommerce.com/wp-json/wccom/inbox-notifications/2.0/notifications.json?locale=en_US

### Testing with Playground

1. [Create a Playground site](https://playground.wordpress.net/?blueprint-url=https://gist.githubusercontent.com/moon0326/ac60ad6389aedf86f8aeded3545faaaf/raw/7af79e663b7eae112c6cfb00ed7f1b8bb5d45301/gistfile1.txt)
2. Go to `Tools -> WCA Test Helper -> Request Redirects`
3. Click on `Edit`
4. Copy the password from secret store 5926 and update it.
5. Go to `Tools -> WCA Test Helper -> Tools` and run `Run wc_admin_daily job`
6. Go to `Tools -> WCA Test Helper -> Admin notes`
7. Click on `List all notes`
8. Confirm the note from the staging has been imported.

### Testing Locally

11. Checkout this branch locally.
12. Cd to `plugins/woocommerce-beta-tester` and run `pnpm run start` to build the changes.
13. Make sure to activate `woocommerce-beta-tester`
14. Open a new terminal session and use wp cli and run the following command.
```
wp option add wc-admin-test-helper-request-redirects '[{"original_endpoint":"https:\/\/woocommerce.com\/wp-json\/wccom\/inbox-notifications\/2.0\/notifications.json?locale=en_US","new_endpoint":"https:\/\/staging.woocommerce.com\/wp-json\/wccom\/inbox-notifications\/2.0\/notifications.json?locale=en_US","username":"woocommerce","password":"","enabled":true}]' --format=json
```
16. Go to `Tools -> WCA Test Helper -> Request Redirects`
17. Click on `Edit` button and update the password from secret store 5926
18. Go to `Tools -> WCA Test Helper -> Tools` and run `Run wc_admin_daily_job`
19. Go to `Tools -> WCA Test Helper -> Admin notes` and click on `List all notes` button.
20. Confirm the note from the staging has been imported.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add 'Request Redirects' tool to the beta tester to help re-routing wp_remote_get calls.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-beta-tester/api/request-redirects/class-wca-test-helper-request-redirects.php [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	 * Create a redirector.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Response
+	 */
+	public static function create_or_update( $request ) {
+		$params = [
+			'original_endpoint' => $request->get_param('original_endpoint'),
+			'new_endpoint' => $request->get_param('new_endpoint'),
+			'username' => $request->get_param('username'),
```
Comment (ilyasfoo) [inline plugins/woocommerce-beta-tester/api/request-redirects/class-wca-test-helper-request-redirects.php:88]: I think it's quite dangerous to have the username and password in plaintext persisted in options. For example in one of our testing scenario, we're using WCCOM's staging site HTTP Auth credentials in a potentially unsecured server, which I think would be a no-no. 

At the very least, we should avoid using staging site directly in an external server and we'll need to add a warning in the UI to warn users about permanently storing the password in plaintext.

Another improvement might be to use a temporary transient as well to ensure credentials (and perhaps the entire configuration for easier management) are not stored longer than it has to.

File: plugins/woocommerce/package.json [hunk 2]
```diff
@@ -695,7 +695,6 @@
 				"node_modules/@woocommerce/e2e-core-tests/CHANGELOG.md",
 				"node_modules/@woocommerce/api/dist/",
 				"node_modules/@woocommerce/admin-e2e-tests/build",
-				"node_modules/@woocommerce/classic-assets/build",
```
Comment (ilyasfoo) [inline plugins/woocommerce/package.json:698]: Is this removal intentional? 🤔

**Review by ilyasfoo — CHANGES_REQUESTED** [review]
Nice work, @moon0326! I have a security concern on storing username and password in plaintext. At the very least, we should put up a warning, for example `Warning: Username and password is stored in plaintext in database. Limit use to testing credentials only and remove after use`. 

This should prevent developers from accidentally storing extra-sensitive credentials in an unsecured server.

Also, I think we could be more clear that the username and password is for HTTP Basic Auth. Perhaps by adding a header above the fields that say `HTTP Basic Auth` 🤔 

image

An issue while testing in local is the options API was disallowed when reading `wc-admin-test-helper-request-redirects` from react:

image

File: plugins/woocommerce-beta-tester/api/request-redirects/class-wca-test-helper-request-redirects.php (hunk 1 above)
Comment (moon0326) [reply plugins/woocommerce-beta-tester/api/request-redirects/class-wca-test-helper-request-redirects.php:88]: That's a good point. I was actually going to suggest to use WP Playground, where everything is stored locally and destroyed when the session is destroyed. However, there is no way for us to limit this plugin to be installed only in Playground.

A temporary transient with a short lifespan sounds good to me. What do you think about storing credentials for an hour? or even 30 mins. That should be long enough for testing. I'm going to think about it a bit further if there are any other safer way.

**Comment by moon0326** [comment]
> An issue while testing in local is the options API was disallowed when reading `wc-admin-test-helper-request-redirects` from react:

Interesting 🤔 I did not get that error from my local or JN sites. I'll look into it.

File: plugins/woocommerce-beta-tester/api/request-redirects/class-wca-test-helper-request-redirects.php (hunk 1 above)
Comment (moon0326) [reply plugins/woocommerce-beta-tester/api/request-redirects/class-wca-test-helper-request-redirects.php:88]: Actually, I think there is a way to get the staging schema without HTTP Basic auth and it's simpler than this approach. I'm going to close this PR for now and re-open with a different tool.

**Comment by moon0326** [comment]
Closing it for now. See https://github.com/woocommerce/woocommerce/pull/48697#discussion_r1648605793
